src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyTests.cs

```diff
@@ -0,0 +1,1005 @@
+                // (4,10): error CS0177: The out parameter 'i' must be assigned to before control leaves the current method
+                //     void M(out readonly int i)
+                Diagnostic(ErrorCode.ERR_ParamUnassigned, "M").WithArguments("i").WithLocation(4, 10));
+        }
+
+        [Fact]
+        public void ReadOnlyReadonlyParameterNotAllowed()
+        {
+            var source = @"
+class C
+{
+    void M(readonly readonly int i)
+    {
+    }
+}
+";
+
+            CreateCompilationWithMscorlib(source, parseOptions: options).VerifyDiagnostics(
+                // (4,21): error CS1107: A parameter can only have one 'readonly' modifier
+                //     void M(readonly readonly int i)
+                Diagnostic(ErrorCode.ERR_DupParamMod, "readonly").WithArguments("readonly").WithLocation(4, 21));
+        }
+
+        [Fact]
+        public void NoAssignmentToReadOnlyParameter()
+        {
+            var source = @"
+class C
+{
+    void M(readonly int i)
+    {
+        i = 0;
+    }
+}
+";
+
+            CreateCompilationWithMscorlib(source, parseOptions: options).VerifyDiagnostics(
+                // (6,9): error CS1656: Cannot assign to 'i' because it is a 'readonly parameter'
+                //         i = 0;
+                Diagnostic(ErrorCode.ERR_AssgReadonlyLocalCause, "i").WithArguments("i", "readonly parameter").WithLocation(6, 9));
+        }
+
+        [Fact]
+        public void NoCompoundAssignmentToReadOnlyParameter()
+        {
+            var source = @"
+class C
+{
+    void M(readonly int i)
+    {
+        i += 0;
+    }
+}
+";
+
+            CreateCompilationWithMscorlib(source, parseOptions: options).VerifyDiagnostics(
+                // (6,9): error CS1656: Cannot assign to 'i' because it is a 'readonly parameter'
+                //         i = 0;
+                Diagnostic(ErrorCode.ERR_AssgReadonlyLocalCause, "i").WithArguments("i", "readonly parameter").WithLocation(6, 9));
+        }
+
+        [Fact]
+        public void NoIncrementToReadOnlyParameter()
+        {
+            var source = @"
+class C
+{
+    void M(readonly int i)
+    {
+        i++;
+    }
+}
+";
+
+            CreateCompilationWithMscorlib(source, parseOptions: options).VerifyDiagnostics(
+                // (6,9): error CS1656: Cannot assign to 'i' because it is a 'readonly parameter'
+                //         i = 0;
+                Diagnostic(ErrorCode.ERR_AssgReadonlyLocalCause, "i").WithArguments("i", "readonly parameter").WithLocation(6, 9));
+        }
+
+        [Fact]
+        public void NoRefPassingOfReadOnlyParameter()
+        {
+            var source = @"
+class C
+{
+    void M(readonly int i)
+    {
+        M2(ref i);
+    }
+
+    void M2(ref int i)
+    {
+    }
+}
+";
+
+            CreateCompilationWithMscorlib(source, parseOptions: options).VerifyDiagnostics(
+                // (6,16): error CS1656: Cannot assign to 'i' because it is a 'readonly parameter'
```

Comment:
Potentially.  But hten shouldn't that also be the case for using/fixed/foreach variables?  Right now we've split the errors, but i don't see why this would be more like 'const' and less like the readonly variables that those statement produce.